hypertrace-trace-enricher/trace-reader/src/main/java/org/hypertrace/trace/reader/attributes/DefaultValueResolver.java

```diff
@@ -72,108 +73,106 @@ private Single<LiteralValue> resolveDefinition(
+  private Optional<LiteralValue> resolveProjection(ValueSource valueSource, Projection projection) {
     switch (projection.getValueCase()) {
       case ATTRIBUTE_ID:
-        return valueSource
-            .executionContext()
-            .wrapSingle(() -> this.attributeClient.get(projection.getAttributeId()))
+        return this.attributeClient
+            .getById(valueSource.requestContext(), projection.getAttributeId())
             .flatMap(attributeMetadata -> this.resolve(valueSource, attributeMetadata));
       case LITERAL:
-        return Single.just(projection.getLiteral());
+        return Optional.of(projection.getLiteral());
       case EXPRESSION:
         return this.resolveExpression(valueSource, projection.getExpression());
       case VALUE_NOT_SET:
       default:
-        return this.buildAndLogErrorLazily("Unrecognized projection type");
+        if (LOGGING_LIMITER.tryAcquire()) {
+          log.error("Unrecognized projection type");
+        }
+        return Optional.empty();
     }
   }
 
-  private Single<LiteralValue> resolveField(
+  private Optional<LiteralValue> resolveField(
       ValueSource valueSource, SourceField sourceField, AttributeKind attributeKind) {
-    return Maybe.fromOptional(valueSource.getSourceField(sourceField, attributeKind))
-        .defaultIfEmpty(LiteralValue.getDefaultInstance());
+    return valueSource.getSourceField(sourceField, attributeKind);
   }
 
-  private Single<LiteralValue> resolveFirstValuePresent(
+  private Optional<LiteralValue> resolveFirstValuePresent(
       ValueSource valueSource,
       AttributeMetadata attributeMetadata,
       AttributeDefinitions definitions) {
 
-    return Observable.fromIterable(definitions.getDefinitionsList())
-        .concatMapMaybe(
-            definition -> this.maybeResolveDefinition(valueSource, attributeMetadata, definition))
-        .first(LiteralValue.getDefaultInstance());
+    return definitions.getDefinitionsList().stream()
+        .map(definition -> this.resolveDefinition(valueSource, attributeMetadata, definition))
+        .filter(Optional::isPresent)
+        .findFirst()
+        .flatMap(Function.identity());
   }
 
-  private Single<LiteralValue> resolveExpression(
+  private Optional<LiteralValue> resolveExpression(
       ValueSource valueSource, ProjectionExpression expression) {
-
-    Single<AttributeProjection> projectionSingle =
-        Maybe.fromOptional(this.attributeProjectionRegistry.getProjection(expression.getOperator()))
-            .switchIfEmpty(
-                buildAndLogErrorLazily(
-                    "Unregistered projection operator: %s", expression.getOperator()));
-    Single<List<LiteralValue>> argumentsSingle =
+    Optional<AttributeProjection> maybeProjection =
+        this.attributeProjectionRegistry.getProjection(expression.getOperator());
+    if (maybeProjection.isEmpty()) {
+      if (LOGGING_LIMITER.tryAcquire()) {
```

Comment:
it may be beneficial to pull out a log function to stop wrapping every statement like this. e.g `throttledLog`